### PR TITLE
Add possibility to skip generating default schemes

### PR DESCRIPTION
**Pull request by olejnjak**
### Short description 📝

In our projects at @AckeeCZ we usually use customized schemes that add pre-build actions which generate inputs for other build phases. Without those inputs, project is not compilable, thus default generated schemes do not make a real sense and only confuse newcomers.

### Solution 📦

Idea is to add possibility to skip generation of those schemes, the most straight-forward solution seems to be adding a property to `ProjectDescription.Project` with default value so it will not break any current API.

### Implementation 👩‍💻👨‍💻

- [x] add new property `shouldGenerateDefaultSchemes` to `Project`
- [x] check this property in `SchemesGenerator` 
- [x] update existing tests

### Questions ❓

I haven't found any tests regarding generated schemes, so right now none tests were added as right now I'm not sure how this feature could be tested and it might be useful if someone could give me a hand with that.

### CODE REVIEW

**Comment by olejnjak** [comment]
Thanks for the feedback 😎

Well I went with the second alternative and all notes should be addressed as tests were updated, documentation has been updated (as I was into that I also fixed the `scheme` property docs).

**Comment by olejnjak** [comment]
Well, thanks @pepibumur. I'm not against any improvement, but honestly I am not that familiar with `TuistConfig` 🤷‍♂ but keeping it backwards compatible should be essential, so current users do not experience sudden loss of their schemes, we should keep in mind that some users will have only those default schemes, so they will lose all of their schemes...I don't think that will be a good experience...

**Comment by kwridan** [comment]
Awesome nice work @olejnjak! Thanks for making the updates.

You can read more about `TuistConfig` here: https://tuist.io/docs/usage/config/ - additionally you can test it out using of the fixtures in the repository:

e.g.

`fixtures/ios_app_with_custom_scheme` which has a Config under `ios_app_with_custom_scheme/Tuist/Config.swift`

@pepibumur using the updated naming @olejnjak adopted with the enum should help alleviate the usage of `disable` :)

**Comment by pepibumur** [comment]
I think we are touching 2 topics here: where the attribute should be defined and what should be the possible values:

- **Where should the attribute be:** This is where I suggested the attribute should be defined in the `Config` file so that the behavior applies consistently across all the projects. One of the things that we are fighting with Tuist is inconsistencies across projects, which is very common in large Xcode projects. By defining the attribute in the `Config.swift` file, all the projects in subdirectories will be generated following the same scheme behavior. That means I'd change the following in your PR:
  - Move the attribute from `Project.swift` to `Config.swift`.
  - Extend the `ProjectGenerator`'s [generate method](https://github.com/tuist/tuist/blob/master/Sources/TuistGenerator/Generator/ProjectGenerator.swift#L76) to take the config as an argument.
  - Adjust the logic [here](https://github.com/tuist/tuist/blob/master/Sources/TuistGenerator/Generator/ProjectGenerator.swift#L132) to generate or not the schemes depending on the configuration.

- **Possible values:** If we take the type that you are suggesting for configuring the scheme generation behavior, these are all the scenarios developers can run into:

| Custom schemes defined | Generation behavior  | Auto-generated | Custom-generated |
|------------------------|----------------------|----------------|------------------|
| Yes                    | defaultAndCustom     | Yes            | Yes              |
| **Yes**                    | **customOnly**           | **No**             | **Yes**              |
| **No**                     | **defaultAndCustom**     | **Yes**            | **No**               |
| No                     | customOnly           | No             | No               |

The simpler is easier for the user, and for us to optimize future functionalities. For that reason I was suggesting to narrow it down to 2 behaviors. Either Tuist is the owner of the schemes, or it's you. When it's you, you take full control of defining your schemes.

If we go down that path, then we don't need a enum, just a boolean, and to make that backwards compatible I was suggesting to call that generation option `. disableAutogeneratedSchemes`. Using `disable` will feel unnatural to the user because they are opting out from a default behavior of Tuist.

**Comment by kwridan** [comment]
Hi @olejnjak  - apologies for the back & forth, just discussed this with @pepibumur regarding the implementation details to help simplify the approach while taking into account the concerns raised.

 - We can go back to a boolean `.disableAutogeneratedSchemes` in `Config.swif`
- It doesn't need to be exposed in `ProjectDescription.Project`  (`Project.swift`) to ensure all projects are configured consistently
 - We can keep this option at the `TuistCore.Project` model as `autoGenerateSchemes` which defaults to `true`
- The option can be propagated from `Config` to `TuistCore.Project` within [`GeneratorModelLoader`](https://github.com/tuist/tuist/blob/master/Sources/TuistLoader/Loaders/GeneratorModelLoader.swift#L79)

This should be an incremental step to help achieve this feature without any major updates to the API while leaving it possible to do so in the future (improving the way configurations are applied to projects). 

Thanks again for submitting this!

**Comment by olejnjak** [comment]
Thanks again for the feedback! 🙂 

I switched the implementation so it uses `Config.swift` as mentioned. Tests and docs should be updated appropriately 😎

So I think all feedback has been addressed 🙂

File: CHANGELOG.md [hunk 1]
```diff
@@ -46,6 +46,7 @@ Please, check out guidelines: https://keepachangelog.com/en/1.0.0/
 - Optimize `TargetNode`'s set operations https://github.com/tuist/tuist/pull/1095 by @kwridan
 - Optimize `BuildPhaseGenerator`'s method of detecting assets and localized files https://github.com/tuist/tuist/pull/1094 by @kwridan
 - Concurrent project generation https://github.com/tuist/tuist/pull/1096 by @kwridan
+- Add possibility to skip generation of default schemes https://github.com/tuist/tuist/pull/1130 by @olejnjak
```
Comment (fortmarek) [inline CHANGELOG.md]: Just a nitpick - I believe this belongs to a `## Next` section

Comment (olejnjak) [reply CHANGELOG.md]: Good catch! Merging master obviously created this issue 👍

File: website/markdown/docs/usage/projectswift.mdx [hunk 2]
```diff
@@ -94,7 +94,7 @@ A `Project.swift` should initialize a variable of type `Project`. It can take an
     {
       name: 'Schemes',
       description:
```
Comment (kwridan) [inline website/markdown/docs/usage/projectswift.mdx]: > minor: _(see The ``[Configuration `generationOptions`](/docs/usage/config/#generationoption)`` to configure this behavior)_

Comment (olejnjak) [reply website/markdown/docs/usage/projectswift.mdx]: Updated 👍

**Comment by kwridan** [comment]
Awesome will merge this in 👍 (will tweak the commit message to reflect the latests changes)

**Pull request merged**